compiler/rustc_mir/src/transform/const_goto.rs

```diff
@@ -0,0 +1,129 @@
+//! This pass optimizes the following sequence
+//! ```rust
+//! bb2: {
+//!     _2 = const true;
+//!     goto -> bb3;
+//! }
+//!
+//! bb3: {
+//!     switchInt(_2) -> [false: bb4, otherwise: bb5];
+//! }
+//! ```
+//! into
+//! ```rust
+//! bb2: {
+//!     _2 = const true;
+//!     goto -> bb5;
+//! }
+//! ```
+
+use crate::transform::{MirPass, MirSource};
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+use rustc_middle::{mir::visit::Visitor, ty::ParamEnv};
+
+use super::simplify::{simplify_cfg, simplify_locals};
+
+pub struct ConstGoto;
+
+impl<'tcx> MirPass<'tcx> for ConstGoto {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, source: MirSource<'tcx>, body: &mut Body<'tcx>) {
+        trace!("Running ConstGoto on {:?}", source);
+        let param_env = tcx.param_env_reveal_all_normalized(source.def_id());
+        let mut opt_finder =
+            ConstGotoOptimizationFinder { tcx, body, optimizations: vec![], param_env };
+        opt_finder.visit_body(body);
+        let should_simplify = !opt_finder.optimizations.is_empty();
+        for opt in opt_finder.optimizations {
+            let terminator = body.basic_blocks_mut()[opt.bb_with_goto].terminator_mut();
+            let new_goto = TerminatorKind::Goto { target: opt.target_to_use_in_goto };
+            debug!("SUCCESS: replacing `{:?}` with `{:?}`", terminator.kind, new_goto);
+            terminator.kind = new_goto;
+        }
+
+        // if we applied optimizations, we potentially have some cfg to cleanup to
+        // make it easier for further passes
+        if should_simplify {
+            simplify_cfg(body);
+            simplify_locals(body, tcx);
+        }
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for ConstGotoOptimizationFinder<'a, 'tcx> {
+    fn visit_terminator(&mut self, terminator: &Terminator<'tcx>, location: Location) {
+        let mut bailer = || {
+            match terminator.kind {
+                TerminatorKind::Goto { target } => {
+                    // We only apply this optimization if the last statement is a const assignment
+                    let last_statement =
+                        self.body.basic_blocks()[location.block].statements.last()?;
+
+                    match &last_statement.kind {
+                        StatementKind::Assign(box (place, Rvalue::Use(op))) => {
+                            let _const = op.constant()?;
```

Comment:
We also need to check that `place` is only ever assigned this value.

Reply:
My thinking was that since this is the last statement, and we check that the target bb has no statements, then place can not be assigned a different value in this sequence if operations.

Reply:
Couldn't there be a different block that also jumps to the switch block? Like this?

```rust
bb2: {
    _2 = const true;
    goto -> bb4;
}

bb3: {
    _2 = const false;
    goto -> bb4;
}

bb4: {
    switchInt(_2) -> [false: bb4, otherwise: bb5];
}
```

Reply:
Your example seems to be an infinite loop, since bb3 jumps to bb4, which then again repeatedly jumps to bb4. Did you intent this loop?

**If you meant false: bb6 in the switchInt of bb4:**

That happens here https://github.com/rust-lang/rust/pull/77486/files#diff-cd637909adce5b1294c12644e9783b7eR19-R31.
I think bb2 and bb3 can be seen as isolated cases.

bb2 will execute the following sequence of operations:
- `_2 = const true;`
- `goto -> bb4;`
- `switchInt(_2) -> [false: bb4, otherwise: bb5];` // since _2 is true, this jumps to bb5.
The optimization will replace `goto -> bb4` with `goto -> bb5` in bb2.

bb3 will execute the following sequence of operations:
- `_2 = const false;`
- `goto -> bb4;`
- `switchInt(_2) -> [false: bb4, otherwise: bb5];` // since _2 is false, this jumps to bb4.
The optimization will replace `goto -> bb4` with `goto -> bb6` in bb2.


In both bb2 and bb3, _2 is not written to in the sequence of operations, which is why I think it is okay.